File: dynamic_language_python/ui/sudokuTableModel.py
from PySide6.QtCore import QAbstractTableModel, QObject, Qt, Property, Signal

from sudoku.sudoku import Sudoku

class SudokuTableModel(QAbstractTableModel):

    messageChanged = Signal()

    def __init__(self, parent = None) -> None:
        super().__init__(parent)
        # 数独
        self.__sudoku = Sudoku()
        # 显示提示
        self.__show_hint = False
        # 反馈信息
        self.__message = str()

    '''
    表格行数
    '''
    def rowCount(self, index):
        return 0 if index.isValid() else self.__sudoku.gridSize()

    '''
    表格列数
    '''
    def columnCount(self, index):
        return 0 if index.isValid() else self.__sudoku.gridSize()

    '''
    表格显示数据
    '''
    def data(self, index, role):
        if not index.isValid() or role != Qt.DisplayRole:
            return None

        # Show all hint numbers
        if self.__show_hint:
            inference_numbers = self.__sudoku.getInference(index.row(), index.column())
            return ' '.join(str(number) for number in inference_numbers)
        
        # Show fixed number
        fixed_number = self.__sudoku.at(index.row(), index.column())
        return "" if fixed_number == 0 else str(fixed_number)

    '''
    解析输入字符串
    '''
    def parse(self, input_string: str) -> bool:
        # 1. Try parsing
        parsing_result = self.__sudoku.parse(input_string)

        # 2. Make some feedback
        self.__message = "Parsing successfully" if parsing_result else "ERROR: Fail to parse"

        # 3. Refresh table
        self.beginResetModel()
        self.endResetModel()

        return parsing_result

    def message(self):
        return self.__message

    def showHint(self):
        return self.__show_hint

    def setShowHint(self, show_hint: bool):
        self.__show_hint = show_hint

        # Update feedback message
        self.__message = "Hints have been showed" if self.__show_hint else "Hints have been hidden"

        # Refresh UI table values
        self.beginResetModel()
        self.endResetModel()

    ''' Bind QML variables '''
    message = Property(str, message, notify=messageChanged)
    show_hint = Property(bool, showHint, setShowHint)

File: static_language_c_plus_plus/src/sudokutablemodel.cpp
#include "sudokutablemodel.h"

SudokuTableModel::SudokuTableModel(QObject* parent /*= nullptr*/)
    : QAbstractTableModel(parent)
{

}

int SudokuTableModel::rowCount(const QModelIndex &parent /*= QModelIndex()*/) const
{
    return parent.isValid() ? 0 : _sudoku.gridSize();
}

int SudokuTableModel::columnCount(const QModelIndex &parent /*= QModelIndex()*/) const
{
    return parent.isValid() ? 0 : _sudoku.gridSize();
}

QVariant SudokuTableModel::data(const QModelIndex &index, int role /*= Qt::DisplayRole*/) const
{
    if (!index.isValid() || role != Qt::DisplayRole)
    {
        return QVariant();
    }

    // 固定值
    auto isFixedNumber = [this, &index](int row, int column){
        return _sudoku.at(row, column) > 0;
    };

    switch (role)
    {
    case Qt::DisplayRole: 
    {
        // Show hint numbers
        if (_show_hint && !isFixedNumber(index.row(), index.column()))
        {
            std::vector<int> cell_hint_numbers = _sudoku.getInference(index.row(), index.column());
            QString hint_numbers_string;
            for(int hint_number : cell_hint_numbers)
                hint_numbers_string += QString::number(hint_number) + " ";
            return hint_numbers_string.mid(0, hint_numbers_string.size() - 1);
        }

        // Show fixed numbers
        int cell_number = _sudoku.at(index.row(), index.column());
        return cell_number == 0 ? QString("") : QString::number(cell_number); 
    }
    }

    return QVariant();
}

bool SudokuTableModel::parse(QString str)
{
    // 1. Parse input string
    bool result = _sudoku.parse(str.toStdString());
    _msg = result ? "Parsing successfully" : "ERROR: Fail to parse!\nPlease check the input string";

    // 2. Refresh table numbers if success
    beginResetModel();
    endResetModel();
    
    return result;
}